src/pip/_internal/req/req_file.py

```diff
@@ -545,7 +545,25 @@ def get_file_content(url: str, session: "PipSession") -> Tuple[str, str]:
     # Assume this is a bare path.
     try:
         with open(url, "rb") as f:
-            content = auto_decode(f.read())
+            raw_content = f.read()
     except OSError as exc:
         raise InstallationError(f"Could not open requirements file: {exc}")
+
+    try:
+        content = auto_decode(raw_content)
+    except UnicodeDecodeError as exc:
+        fallback_encoding = "utf-8"
+        # don't try an decode again if we know it will fail
+        if exc.encoding == fallback_encoding:
+            raise
+
+        logging.warning(
+            "unable to decode data from %s with encoding %s, "
+            "falling back to encoding %s",
+            url,
+            exc.encoding,
+            fallback_encoding,
+        )
+        content = raw_content.decode(fallback_encoding)
```

Comment:
It may be a good idea to use `error="backslashreplace"` here. Most of the time, the offending bytes would just be a part of a comment anyway and would not make a difference.

Reply:
> It may be a good idea to use `error="backslashreplace"` here. Most of the time, the offending bytes would just be a part of a comment anyway and would not make a difference.

I've been hesitating with this a bit, specifically I'm wondering if this could be abused for nefarious purposes where the contents of the file you 'see' (not well defined, since this is the case where the data won't fully decode) isn't the same contents that pip will process. Though I'm having a hard time finding a vulnerable use-case (e.g injecting an extra element or a adding a `.` to a domain name in a requirement URL)

Reply:
Generally, we're trying to enforce compliance with the documentation, so I don't think we should introduce another hack to support technically mal-encoded requirement files.